x/protocolpool/keeper/keeper.go

```diff
@@ -67,3 +88,124 @@ func (k Keeper) GetCommunityPool(ctx context.Context) (sdk.Coins, error) {
+	if budget.TranchesLeft == 0 && budget.ClaimedAmount != nil {
+		// check that claimed amount is equal to total budget
+		if budget.ClaimedAmount.Equal(budget.TotalBudget) {
+			// remove the entry of budget ended recipient
+			if err := k.BudgetProposal.Remove(ctx, recipient); err != nil {
+				return sdk.Coin{}, err
+			}
+			// Return the end of the budget
+			return sdk.Coin{}, fmt.Errorf("budget ended for recipient: %s", recipient.String())
+		}
+	}
+
+	currentTime := sdkCtx.BlockTime().Unix()
+	startTime := budget.StartTime
+
+	// Check if the start time is reached
+	if uint64(currentTime) < startTime {
+		return sdk.Coin{}, fmt.Errorf("distribution has not started yet")
+	}
+
+	if budget.NextClaimFrom == 0 {
+		budget.NextClaimFrom = budget.StartTime
+	}
+
+	if budget.TranchesLeft == 0 && budget.ClaimedAmount == nil {
+		budget.TranchesLeft = budget.Tranches
+		zeroCoin := sdk.NewCoin(budget.TotalBudget.Denom, math.ZeroInt())
+		budget.ClaimedAmount = &zeroCoin
+	}
+
+	return k.calculateClaimableFunds(ctx, recipient, budget, currentTime)
+}
+
+func (k Keeper) calculateClaimableFunds(ctx context.Context, recipient sdk.AccAddress, budget types.Budget, currentTime int64) (amount sdk.Coin, err error) {
+	// Calculate the time elapsed since the last claim time
+	timeElapsed := uint64(currentTime) - budget.NextClaimFrom
+
+	// Check the time elapsed has passed period length
+	if timeElapsed < budget.Period {
+		return sdk.Coin{}, fmt.Errorf("budget period has not passed yet")
+	}
+
+	// Calculate how many periods have passed
+	periodsPassed := timeElapsed / budget.Period
+
+	// Calculate the amount to distribute for all passed periods
+	coinsToDistribute := math.NewInt(int64(periodsPassed)).Mul(budget.TotalBudget.Amount.QuoRaw(int64(budget.Tranches)))
+	amount = sdk.NewCoin(budget.TotalBudget.Denom, coinsToDistribute)
+
+	// update the budget's remaining tranches
+	budget.TranchesLeft -= periodsPassed
+
+	// update the ClaimedAmount
+	claimedAmount := budget.ClaimedAmount.Add(amount)
+	budget.ClaimedAmount = &claimedAmount
+
+	// Update the last claim time for the budget
+	budget.NextClaimFrom += budget.Period
+
+	k.Logger(ctx).Debug(fmt.Sprintf("Processing budget for recipient: %s. Amount: %s", budget.RecipientAddress, coinsToDistribute.String()))
+
+	// Save the updated budget in the state
+	if err := k.BudgetProposal.Set(ctx, recipient, budget); err != nil {
+		return sdk.Coin{}, fmt.Errorf("error while updating the budget for recipient %s", budget.RecipientAddress)
+	}
+
+	return amount, nil
+}
+
+func (k Keeper) validateBudgetProposal(ctx context.Context, bp types.MsgSubmitBudgetProposal) error {
+	if bp.TotalBudget.IsZero() {
+		return fmt.Errorf("invalid budget proposal: total budget cannot be zero")
+	}
+
+	if err := validateAmount(sdk.NewCoins(*bp.TotalBudget)); err != nil {
+		return fmt.Errorf("invalid budget proposal: %w", err)
+	}
+
+	if int64(bp.StartTime) <= sdk.UnwrapSDKContext(ctx).BlockTime().Unix() {
```

Comment:
should we remove this check? cc @julienrbrt 

Talked with likhita and this could make submitting a proposal harder because you need to calculate when the proposal will end

Reply:
Yes, my suggestion was let it be 0 if wanted, and set it to the current block time when the proposal passed.
So it would be:

```go
if int64(bp.StartTime)== 0 {
	bp.StartTime =  sdk.UnwrapSDKContext(ctx).BlockTime().Unix()
}
```